### DISCUSSION TITLE
New port almost ready

### DISCUSSION THREAD

**Discussion by jesperoman** [sqids/sqids-scala#1] [Show and tell]
I have a scala version of this [here](https://github.com/jesperoman/sqids-scala)
It has some polishing, publishing of artifacts and stuff left, but let me know if you want it pushed here.
Should I fork and make a pull request?

**Comment by 4kimov** [comment]
Oh cool. How about I give you maintainer rights to this repo and you push your code here?

**Reply by jesperoman** [reply]
Yeah, sure!
Works fine for me!

**Comment by 4kimov** [comment]
Cool, invite sent.

A few notes:
- If you'd like to use one of the existing READMEs, you can copy these and just replace examples: [Go](https://github.com/sqids/sqids-go), [Rust](https://github.com/sqids/sqids-rust), [Python](https://github.com/sqids/sqids-python)
- I can see you got most tests there which is great, perhaps [just this one missing](https://github.com/sqids/sqids-spec/commit/f52b57836b0463097018f984f853b284e50a5ce4)?
- If you need to store secrets in the repo (for testing or publishing package), please let me know
- When you're ready with the repo + package, just let me know and I'll update the site

And of course, I'm here if you have questions about anything 👍

**Reply by jesperoman** [reply]
Cool! I will get started transferring the repo as soon as possible, and complete missing test(s)

**Reply by jesperoman** [reply]
I think most of the stuff is done now, including running tests with github actions, will add badges to the readme as well.
About publishing of artifacts, I've signed up to [OSSRH](https://central.sonatype.org/publish/publish-guide/) and issued a ticket to be able to publish the library to maven central.

I figured I'll use the group id `org.sqids` but I think an owner of the domain might need to add a TXT record to the domain to prove ownership later in the process, let me know if you're able to help me with that!

**Reply by 4kimov** [reply]
No problem on the DNS record, just let me know. I'll update the site when you're done with all 👍

**Reply by jesperoman** [reply]
Cool, thanks!
Here's the request to create a repository with sonatype:
https://issues.sonatype.org/browse/OSSRH-94320

Apparently you need to add a TXT record with the text: `OSSRH-94320` to the domain!

**Reply by jesperoman** [reply]
Another question, is this the best channel of communication of sqids implementations, or do you have any other preferences, twitter, gitter or something?

**Reply by 4kimov** [reply]
- Cool, thanks for linking. Added TXT record.
- For now there's no group chat for Sqids, because I haven't seen the demand yet, but maybe I should make one... So atm Github is probably best, in case anyone else has similar questions

**Reply by jesperoman** [reply]
Yep, cool, preparing to be able to publish artifacts to public repos.

**Reply by jesperoman** [reply]
Ok, now I need to add some repository secrets to this repo to be able to publish artifacts based on git tags.
How do I send those to you?

**Reply by 4kimov** [reply]
Just bumped you up to an admin, you can now add secrets directly.

**Reply by jesperoman** [reply]
Cool, thanks!

**Reply by jesperoman** [reply]
Now I think there's actually a version 0.1.0 out there in the wild!

**Reply by 4kimov** [reply]
Cool, good job! 💪  Two questions for you:

1. Can we have the release badge link to wherever people can find the package?
2. Should I update the site?

**Reply by jesperoman** [reply]
Hmm, yeah, maybe we should, I just created a [new PR](https://github.com/sqids/sqids-scala/pull/2) linking the release badge to the github release.
Thing is, it's pretty uncommon in the scala community to manually download an artifact and include in some classpath, usually it's specified in the build file, but I'll see what possibilities we have.
One way could be to link it here: https://central.sonatype.com/artifact/org.sqids/sqids_2.13

**Reply by jesperoman** [reply]
Now I think I found a nice link, check out the [new readme](https://github.com/sqids/sqids-scala/blob/links_in_readme/README.md), and feel free to approve [the PR](https://github.com/sqids/sqids-scala/pull/2) 😉

**Reply by 4kimov** [reply]
Nice, ok cool. A few follow-ups:

1. Website has been [updated](https://sqids.org/scala). Anything you'd like me to adjust in examples?
2. I merged your PR 👍 The badges look nice and the link to Scaladex is great
3. Just curious: is it Scala convention to name the function "encodeUnsafeString" (as opposed to just "encode")?
4. Thank you for all the great work! 💪

**Reply by jesperoman** [reply]
Cool!
About encodeUnsafeString, there is an encode(numbers) function as well, but it returns an "Either" instead of throwing exceptions, so that's the `safe` function. 
That's a pretty common pattern in scala, as it works in the haskell version.

Maybe I should change examples to use that instead, to make it clear how it works?

**Reply by jesperoman** [reply]
Hmm, thinking of it, the only reason for returning an Either (or throw an Exception) when encoding is the blocklist handling, there is a slight risk of that breaking, if there is >= Int.MaxValue iterations of an id which all contains a blocked word, that seems very unlikely, maybe we could live with that throwing an exception instead of having to return an either.
That's a breaking change on next version, but maybe we can live with that, don't think there is heavy usage of this yet 😅

**Reply by jesperoman** [reply]
Oh wait, there is also the possibility to try and encode numbers outside of the allowed (like negative numbers), hmm, I'll think about it, will probably have a new release in a day or two.

**Reply by jesperoman** [reply]
Also thinking about using `Long`'s instead of `Int`'s, Actually already have a branch with that implementation, since javascript version have a much larger maxValue than scala, it could produce id's not decodable by the current scala version, possibly bad.
It's a pretty significant difference:
```scala
scala> Int.MaxValue
val res0: Int = 2147483647

scala> Long.MaxValue
val res1: Long = 9223372036854775807
```

**Reply by 4kimov** [reply]
I see, so according to spec, `encode` function is capable of throwing an error (during an overflow), and `decode` function is not (it will just return an empty array if anything fails). I don't think we should estimate the likelihood of `encode` hitting the blocklist limit because we don't control the constructor/input params (the user does), so it should def be designed with the expectation that it can fail easily (imagine the user giving a giant blocklist of all possible values). I leave it up to you how to design it in Scala (via Either or exceptions).

Regarding int vs long, personally I just use the largest unsigned integer the language supports (eg: `u64` for Rust, `uint64` for Go). I'm guessing you're not using only `Long` everywhere because of performance? If that's the case, I'd just suggest to do whatever the Scala community recommends regarding safely + efficiently handling large numbers.

**Reply by jesperoman** [reply]
That settles it, Long in scala is basically a 64 bit integer the scala equivalent of rusts u64.
From the scaladoc:
_`Long`, a 64-bit signed integer (equivalent to Java's `long` primitive type)_
Unsigned integers dont exist in the jvm world.

When handling lengths of strings and such, of course I'm using regular ints (32 bit) since that's the default.

**Reply by 4kimov** [reply]
Ok, sounds good. Then the only thing to handle is manual checks for negative numbers (I believe the spec repo shows examples of those, in the library and in tests).

**Reply by jesperoman** [reply]
Yes there is already [checks for that](https://github.com/sqids/sqids-scala/blob/1d80dd1acef9030601d7f6eea5b03f380d320978/src/main/scala/sqid/Sqids.scala#L103-L108) and [tests as well](https://github.com/sqids/sqids-scala/blob/1d80dd1acef9030601d7f6eea5b03f380d320978/src/test/scala/sqids/SqidsSuite_encoding.scala#L107-L121)
